apps/cmp.c

```diff
@@ -0,0 +1,3256 @@
+            if (srvcert == NULL)
+                /*
+                 * opt_otherpass is needed in case
+                 * opt_srvcert is an encrypted PKCS#12 file
+                 */
+                goto err;
+            if (!OSSL_CMP_CTX_set1_srvCert(ctx, srvcert)) {
+                X509_free(srvcert);
+                goto oom;
+            }
+            X509_free(srvcert);
+            if ((ts = X509_STORE_new()) == NULL)
+                goto oom;
+        }
+        if (opt_trusted != NULL
+                && (ts = load_certstore(opt_trusted, "trusted certificates"))
+            == NULL)
+            goto err;
+        if (!set1_store_parameters(ts) /* also copies vpm */
+                /*
+                 * clear any expected host/ip/email address;
+                 * opt_expect_sender is used instead
+                 */
+                || !truststore_set_host_etc(ts, NULL)
+                || !OSSL_CMP_CTX_set0_trustedStore(ctx, ts)) {
+            X509_STORE_free(ts);
+            goto oom;
+        }
+    }
+
+    if (opt_ignore_keyusage)
+        (void)OSSL_CMP_CTX_set_option(ctx, OSSL_CMP_OPT_IGNORE_KEYUSAGE, 1);
+
+    if (opt_unprotected_errors)
+        (void)OSSL_CMP_CTX_set_option(ctx, OSSL_CMP_OPT_UNPROTECTED_ERRORS, 1);
+
+    if (opt_out_trusted != NULL) { /* for use in OSSL_CMP_certConf_cb() */
+        X509_VERIFY_PARAM *out_vpm = NULL;
+        X509_STORE *out_trusted = load_certstore(opt_out_trusted,
+                                                 "trusted certs for verifying newly enrolled cert");
+        if (out_trusted == NULL)
+            goto err;
+        /* any -verify_hostname, -verify_ip, and -verify_email apply here */
+        if (!set1_store_parameters(out_trusted))
+            goto oom;
+        /* ignore any -attime here, new certs are current anyway */
+        out_vpm = X509_STORE_get0_param(out_trusted);
+        X509_VERIFY_PARAM_clear_flags(out_vpm, X509_V_FLAG_USE_CHECK_TIME);
+
+        (void)OSSL_CMP_CTX_set_certConf_cb_arg(ctx, out_trusted);
+    }
+
+    if (opt_disable_confirm)
+        (void)OSSL_CMP_CTX_set_option(ctx, OSSL_CMP_OPT_DISABLE_CONFIRM, 1);
+
+    if (opt_implicit_confirm)
+        (void)OSSL_CMP_CTX_set_option(ctx, OSSL_CMP_OPT_IMPLICIT_CONFIRM, 1);
+
+    (void)OSSL_CMP_CTX_set_certConf_cb(ctx, OSSL_CMP_certConf_cb);
+
+    return 1;
+
+ oom:
+    CMP_err("out of memory");
+ err:
+    return 0;
+}
+
+#ifndef OPENSSL_NO_SOCK
+/*
+ * set up ssl_ctx for the OSSL_CMP_CTX based on options from config file/CLI.
+ * Returns pointer on success, NULL on error
+ */
+static SSL_CTX *setup_ssl_ctx(OSSL_CMP_CTX *ctx, ENGINE *e)
+{
+    STACK_OF(X509) *untrusted_certs = OSSL_CMP_CTX_get0_untrusted_certs(ctx);
+    EVP_PKEY *pkey = NULL;
+    X509_STORE *trust_store = NULL;
+    SSL_CTX *ssl_ctx;
+    int i;
+
+    ssl_ctx = SSL_CTX_new(TLS_client_method());
+    if (ssl_ctx == NULL)
+        goto err;
```

Comment:
nit : directly `return 0;` , as there is nothing to clean yet.

Reply:
yep, done